File: app/src/main/java/com/example/myapplication/activities/animepage/watch_anime_mvp/WatchAnimeEpisodeActivity.java
package com.example.myapplication.activities.animepage.watch_anime_mvp;

import androidx.appcompat.app.AppCompatActivity;

import android.annotation.SuppressLint;
import android.app.ProgressDialog;
import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.os.Handler;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.webkit.JavascriptInterface;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.RelativeLayout;
import android.widget.Toast;

import com.example.myapplication.activities.animepage.anime_detail_mvp.AnimeDetailActivity;
import com.example.myapplication.R;
import com.example.myapplication.models.animemodels.VideoStreamResultModel;
import com.example.myapplication.databinding.ActivityWatchAnimeEpisodeBinding;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.select.Elements;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class WatchAnimeEpisodeActivity extends AppCompatActivity implements WatchAnimeEpisodeInterface {

    private ActivityWatchAnimeEpisodeBinding animeEpisodeBinding;
    private VideoStreamResultModel videoStreamResultModel = new VideoStreamResultModel();
    private ProgressDialog progressDialog;
    private String nowEpisodeNumber = "", animeType = "", animeStatus = "", animeThumb = "";
    private WatchAnimeEpisodePresenter episodePresenter = new WatchAnimeEpisodePresenter(this);
    private boolean isShowAtasBawah = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        animeEpisodeBinding = ActivityWatchAnimeEpisodeBinding.inflate(getLayoutInflater());
        setContentView(animeEpisodeBinding.getRoot());
        getWindow().getDecorView().setSystemUiVisibility(0x10);
        initUI();
        initEvent();
    }

    @SuppressLint("ClickableViewAccessibility")
    private void initEvent() {
        animeEpisodeBinding.animeInfoButton.setOnClickListener(v -> {
            if (videoStreamResultModel.getAnimeDetailURL() == null || videoStreamResultModel.getEpisodeTitle() == null) {
                Log.e("URL NULL", "YES");
            } else {
                Intent intent = new Intent(WatchAnimeEpisodeActivity.this, AnimeDetailActivity.class);
                intent.putExtra("animeDetailURL", videoStreamResultModel.getAnimeDetailURL());
                intent.putExtra("animeDetailTitle", videoStreamResultModel.getEpisodeTitle());
                intent.putExtra("animeDetailType", animeType);
                intent.putExtra("animeDetailStatus", animeStatus);
                intent.putExtra("animeDetailThumb", animeThumb);
                startActivity(intent);
                finish();
            }
        });
        animeEpisodeBinding.prevEpisodeButton.setOnClickListener(view -> getWatchAnimeDataFromWebView(videoStreamResultModel.getVideoPrevUrl()));
        animeEpisodeBinding.nextEpisodeButton.setOnClickListener(view -> getWatchAnimeDataFromWebView(videoStreamResultModel.getVideoNextUrl()));
        animeEpisodeBinding.buttonFull.setOnClickListener(v -> {
            isShowAtasBawah = !isShowAtasBawah;
            isAtasBawahShow(isShowAtasBawah);
        });
    }

    @SuppressLint("SetJavaScriptEnabled")
    private void initUI() {
        progressDialog = new ProgressDialog(WatchAnimeEpisodeActivity.this);
        progressDialog.setCanceledOnTouchOutside(false);
        progressDialog.setCancelable(false);
        progressDialog.setMessage("Be patient please onii-chan, it just take less than a minute :3");
        animeEpisodeBinding.webViewWatchAnime.getSettings().setUserAgentString("Mozilla/5.0 (Windows NT 6.1; rv:13.0) Gecko/20100101 Firefox/12");
        animeEpisodeBinding.webViewWatchAnime.getSettings().setJavaScriptEnabled(true);
        animeEpisodeBinding.webViewWatchAnime.getSettings().setLoadWithOverviewMode(true);
        animeEpisodeBinding.webViewWatchAnime.getSettings().setUseWideViewPort(false);
        String episodeURL = getIntent().getStringExtra("animeEpisodeToWatch");
        String episodeTitle = getIntent().getStringExtra("animeEpisodeTitle");
        animeType = getIntent().getStringExtra("animeEpisodeType");
        animeStatus = getIntent().getStringExtra("animeEpisodeStatus");
        animeThumb = getIntent().getStringExtra("animeEpisodeThumb");
        if (episodeURL != null && episodeTitle != null && animeType != null) {
            if (animeType.equalsIgnoreCase(getResources().getString(R.string.series_string)) ||
                    animeType.contains(getResources().getString(R.string.series_string))) {
                animeEpisodeBinding.linearAbove.setBackgroundColor(getResources().getColor(R.color.blue_series_color));
                animeEpisodeBinding.linearBelow.setBackgroundColor(getResources().getColor(R.color.blue_series_color));
            } else if (animeType.equalsIgnoreCase(getResources().getString(R.string.ona_string)) ||
                    animeType.contains(getResources().getString(R.string.ona_string))) {
                animeEpisodeBinding.linearAbove.setBackgroundColor(getResources().getColor(R.color.purple_series_color));
                animeEpisodeBinding.linearBelow.setBackgroundColor(getResources().getColor(R.color.purple_series_color));
            } else if (animeType.equalsIgnoreCase(getResources().getString(R.string.movie_string)) ||
                    animeType.contains(getResources().getString(R.string.movie_string_lower))) {
                animeEpisodeBinding.linearAbove.setBackgroundColor(getResources().getColor(R.color.green_series_color));
                animeEpisodeBinding.linearBelow.setBackgroundColor(getResources().getColor(R.color.green_series_color));
                animeEpisodeBinding.nextEpisodeButton.setVisibility(View.GONE);
                animeEpisodeBinding.prevEpisodeButton.setVisibility(View.GONE);
            } else if (animeType.equalsIgnoreCase(getResources().getString(R.string.la_string)) ||
                    animeType.contains(getResources().getString(R.string.la_string))) {
                animeEpisodeBinding.linearAbove.setBackgroundColor(getResources().getColor(R.color.red_series_color));
                animeEpisodeBinding.linearBelow.setBackgroundColor(getResources().getColor(R.color.red_series_color));
            } else if (animeType.equalsIgnoreCase(getResources().getString(R.string.special_string)) ||
                    animeType.contains(getResources().getString(R.string.special_string_lower))) {
                animeEpisodeBinding.linearAbove.setBackgroundColor(getResources().getColor(R.color.orange_series_color));
                animeEpisodeBinding.linearBelow.setBackgroundColor(getResources().getColor(R.color.orange_series_color));
            } else if (animeType.equalsIgnoreCase(getResources().getString(R.string.ova_string)) ||
                    animeType.contains(getResources().getString(R.string.ova_string))) {
                animeEpisodeBinding.linearAbove.setBackgroundColor(getResources().getColor(R.color.pink_series_color));
                animeEpisodeBinding.linearBelow.setBackgroundColor(getResources().getColor(R.color.pink_series_color));
            }
            animeEpisodeBinding.textAnimeTitleWatch.setText(episodeTitle);
            getWatchAnimeDataFromWebView(episodeURL);
        } else {
            Log.e("nowURL", "gak ada");
        }
    }

    @SuppressLint("SetJavaScriptEnabled")
    private void getWatchAnimeDataFromWebView(String episodeURL) {
        progressDialog.show();
        List<String> splitList = Arrays.asList(episodeURL.split("-"));
        nowEpisodeNumber = splitList.get(splitList.indexOf("episode") + 1);
        animeEpisodeBinding.webViewWatchAnimeBg.getSettings().setJavaScriptEnabled(true);
        animeEpisodeBinding.webViewWatchAnimeBg.addJavascriptInterface(new LoadListener(), "HTMLOUT");
        animeEpisodeBinding.webViewWatchAnimeBg.loadUrl(episodeURL);
        if (animeEpisodeBinding.webViewWatchAnimeBg.isShown()) {
            animeEpisodeBinding.webViewWatchAnimeBg.reload();
        }
        animeEpisodeBinding.webViewWatchAnimeBg.setWebViewClient(new WebViewClient() {
            @Override
            public boolean shouldOverrideUrlLoading(WebView view, String url) {

                return true;
            }

            @Override
            public void onPageStarted(WebView view, String url,
                                      Bitmap favicon) {
            }

            public void onPageFinished(WebView view, String url) {
                new Handler().postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        view.loadUrl("javascript:window.HTMLOUT.processHTML('<html>'+document.getElementsByTagName('html')[0].innerHTML+'</html>');");
                    }
                }, 5000);
            }

        });
    }

    public void showToast(String getURLFromElementSrc, String message) {
        runOnUiThread(() -> {
            Toast.makeText(WatchAnimeEpisodeActivity.this, message, Toast.LENGTH_SHORT).show();
            if (!getURLFromElementSrc.isEmpty()) {
                episodePresenter.getEpisodeToWatchData(nowEpisodeNumber, message);
            } else {
                animeEpisodeBinding.webViewWatchAnimeBg.reload();
                Toast.makeText(WatchAnimeEpisodeActivity.this, "Failed!", Toast.LENGTH_SHORT).show();
            }
        });
    }

    class LoadListener {
        public String content = "";

        @JavascriptInterface
        public void processHTML(String html) {
            Document doc = Jsoup.parse(html);
            Elements getVideoEmbedURL = doc.getElementsByClass("playeriframe");
            String getURLFromElementSrc = getVideoEmbedURL.attr("src");
            Log.e("result", getURLFromElementSrc);
            nextToAct(getURLFromElementSrc, html);
        }

        private void nextToAct(String getURLFromElementSrc, String content) {
            showToast(getURLFromElementSrc, content);
        }
    }

    private void isAtasBawahShow(boolean showStatus) {
        if (showStatus) {
            RelativeLayout.LayoutParams layoutParams = (RelativeLayout.LayoutParams) animeEpisodeBinding.webViewWatchAnime.getLayoutParams();
            layoutParams.width = RelativeLayout.LayoutParams.MATCH_PARENT;
            layoutParams.height = RelativeLayout.LayoutParams.MATCH_PARENT;
            animeEpisodeBinding.webViewWatchAnime.setLayoutParams(layoutParams);
            animeEpisodeBinding.linearAbove.setVisibility(View.GONE);
            animeEpisodeBinding.linearBelow.setVisibility(View.GONE);
        } else {
            RelativeLayout.LayoutParams layoutParams = (RelativeLayout.LayoutParams) animeEpisodeBinding.webViewWatchAnime.getLayoutParams();
            layoutParams.width = RelativeLayout.LayoutParams.MATCH_PARENT;
            layoutParams.height = (int) getResources().getDimension(R.dimen._300sdp);
            animeEpisodeBinding.webViewWatchAnime.setLayoutParams(layoutParams);
            animeEpisodeBinding.linearAbove.setVisibility(View.VISIBLE);
            animeEpisodeBinding.linearBelow.setVisibility(View.VISIBLE);
        }
    }

    private void parseHtmlToViewableContent(VideoStreamResultModel result) {
        videoStreamResultModel = result;
        if (result != null) {
            animeEpisodeBinding.textAnimeTitleWatch.setText(result.getEpisodeTitle());
            if (result.getVideoPrevUrl() != null && !result.getVideoPrevUrl().isEmpty()) {
                animeEpisodeBinding.prevEpisodeButton.setVisibility(View.VISIBLE);
            } else {
                animeEpisodeBinding.prevEpisodeButton.setVisibility(View.GONE);
            }

            if (result.getVideoNextUrl() != null && !result.getVideoNextUrl().isEmpty()) {
                animeEpisodeBinding.nextEpisodeButton.setVisibility(View.VISIBLE);
            } else {
                animeEpisodeBinding.nextEpisodeButton.setVisibility(View.GONE);
            }
            if (result.getVideoUrl() != null && !result.getVideoUrl().isEmpty()) {
                DisplayMetrics displayMetrics = getResources().getDisplayMetrics();
                double screenWidth = displayMetrics.widthPixels / displayMetrics.density;
                double screenHeight = displayMetrics.heightPixels / displayMetrics.density;
//                String imageURLModify = "<html><body style=\"margin: 0;\"><iframe width=\"" + screenWidth + "\" height=\"" + screenHeight + "\" src=\"" + result.getVideoUrl() + "\"></iframe></body></html>";
//                String imageURLModify = "<html><body style=\"margin: 0; padding: 0\"><iframe width=\"100%\" height=\"100%\" src=\"" + result.getVideoUrl() + "\" allowfullscreen=\"allowfullscreen\"></iframe></body></html>";
//                animeEpisodeBinding.webViewWatchAnime.loadData(imageURLModify, "text/html", "UTF-8");
                if (animeEpisodeBinding.webViewWatchAnime.isShown()) {
                    animeEpisodeBinding.webViewWatchAnime.reload();
                }
                animeEpisodeBinding.webViewWatchAnime.loadUrl(result.getVideoUrl());
            }

        }

    }

    @Override
    public void onWindowFocusChanged(boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);
        if (hasFocus) {
            hideSystemUI();
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (progressDialog != null) {
            progressDialog.dismiss();
            progressDialog = null;
        }
    }

    private void hideSystemUI() {
        // Enables regular immersive mode.
        // For "lean back" mode, remove SYSTEM_UI_FLAG_IMMERSIVE.
        // Or for "sticky immersive," replace it with SYSTEM_UI_FLAG_IMMERSIVE_STICKY
        View decorView = getWindow().getDecorView();
        decorView.setSystemUiVisibility(
                View.SYSTEM_UI_FLAG_IMMERSIVE
                        // Set the content to appear under the system bars so that the
                        // content doesn't resize when the system bars hide and show.
                        | View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                        | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                        | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                        // Hide the nav bar and status bar
                        | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                        | View.SYSTEM_UI_FLAG_FULLSCREEN);
    }

    @Override
    public void onGetWatchAnimeEpisodeDataSuccess(VideoStreamResultModel watchHTMLResult) {
        progressDialog.dismiss();
        parseHtmlToViewableContent(watchHTMLResult);
    }

    @Override
    public void onGetWatchAnimeEpisodeDataFailed() {
        progressDialog.dismiss();
        Toast.makeText(WatchAnimeEpisodeActivity.this, "Your internet connection is worse than your face onii-chan :3", Toast.LENGTH_SHORT).show();
    }
}

File: app/src/main/java/com/example/myapplication/activities/animepage/watch_anime_mvp/WatchAnimeEpisodeActivity.kt
package com.example.myapplication.activities.animepage.watch_anime_mvp

import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil

import android.annotation.SuppressLint
import android.app.ProgressDialog
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import android.webkit.WebChromeClient
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.Toast

import com.example.myapplication.activities.animepage.anime_detail_mvp.AnimeDetailActivity
import com.example.myapplication.R
import com.example.myapplication.models.animemodels.VideoStreamResultModel
import com.example.myapplication.databinding.ActivityWatchAnimeEpisodeBinding
import com.google.gson.Gson

import org.jsoup.Jsoup
import java.util.ArrayList

class WatchAnimeEpisodeActivity : AppCompatActivity(), WatchAnimeEpisodeInterface {

    private var animeEpisodeBinding: ActivityWatchAnimeEpisodeBinding? = null
    private val videoStreamResultModel = VideoStreamResultModel()
    private var progressDialog: ProgressDialog? = null
    private var nowEpisodeNumber: String? = null
    private var nextEpisodeNumber: String? = null
    private var previousEpisodeNumber: String? = null
    private var nextURL: String? = null
    private var prevURL: String? = null
    private var episodeTitle: String? = null
    private val episodePresenter = WatchAnimeEpisodePresenter(this)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        animeEpisodeBinding = DataBindingUtil.setContentView(this, R.layout.activity_watch_anime_episode)
        initUI()
        initEvent()
    }

    private fun initEvent() {
        animeEpisodeBinding!!.animeInfoButton.setOnClickListener {
            if (videoStreamResultModel.animeDetailURL == null || videoStreamResultModel.episodeTitle == null) {
                Log.e("URL NULL", "YES")
            } else {
                val intent = Intent(this@WatchAnimeEpisodeActivity, AnimeDetailActivity::class.java)
                intent.putExtra("animeDetailURL", videoStreamResultModel.animeDetailURL)
                intent.putExtra("animeDetailTitle", videoStreamResultModel.episodeTitle)
                intent.putExtra("animeDetailType", videoStreamResultModel.animeType)
                intent.putExtra("animeDetailStatus", videoStreamResultModel.animeStatus)
                intent.putExtra("animeDetailThumb", videoStreamResultModel.animeThumb)
                startActivity(intent)
                finish()
            }
        }
        animeEpisodeBinding!!.prevEpisodeButton.setOnClickListener { getAnimeWatchData(prevURL!!) }
        animeEpisodeBinding!!.nextEpisodeButton.setOnClickListener { getAnimeWatchData(nextURL!!) }
    }

    @SuppressLint("SetJavaScriptEnabled")
    private fun initUI() {
        progressDialog = ProgressDialog(this@WatchAnimeEpisodeActivity)
        progressDialog!!.setCanceledOnTouchOutside(false)
        progressDialog!!.setCancelable(false)
        progressDialog!!.setMessage("Be patient please onii-chan, it just take less than a minute :3")
        animeEpisodeBinding!!.webViewWatchAnime.webViewClient = object : WebViewClient() {
            override fun shouldOverrideUrlLoading(webView: WebView, request: WebResourceRequest): Boolean {
                webView.loadUrl(request.url.toString())
                return true
            }
        }
        animeEpisodeBinding!!.webViewWatchAnime.settings.javaScriptEnabled = true
        animeEpisodeBinding!!.webViewWatchAnime.webChromeClient = WebChromeClient()
        val episodeURL = intent.getStringExtra("animeEpisodeToWatch")
        episodeTitle = intent.getStringExtra("animeEpisodeTitle")
        val episodeType = intent.getStringExtra("animeEpisodeType")
        val episodeStatus = intent.getStringExtra("animeEpisodeStatus")
        val episodeThumb = intent.getStringExtra("animeEpisodeThumb")
        if (episodeURL != null && episodeTitle != null && episodeType != null) {
            if (episodeType.equals(resources.getString(R.string.series_string), ignoreCase = true) || episodeType.contains(resources.getString(R.string.series_string))) {
                animeEpisodeBinding!!.linearAbove.setBackgroundColor(resources.getColor(R.color.blue_series_color))
                animeEpisodeBinding!!.linearBelow.setBackgroundColor(resources.getColor(R.color.blue_series_color))
            } else if (episodeType.equals(resources.getString(R.string.ona_string), ignoreCase = true) || episodeType.contains(resources.getString(R.string.ona_string))) {
                animeEpisodeBinding!!.linearAbove.setBackgroundColor(resources.getColor(R.color.purple_series_color))
                animeEpisodeBinding!!.linearBelow.setBackgroundColor(resources.getColor(R.color.purple_series_color))
            } else if (episodeType.equals(resources.getString(R.string.movie_string), ignoreCase = true) || episodeType.contains(resources.getString(R.string.movie_string_lower))) {
                animeEpisodeBinding!!.linearAbove.setBackgroundColor(resources.getColor(R.color.green_series_color))
                animeEpisodeBinding!!.linearBelow.setBackgroundColor(resources.getColor(R.color.green_series_color))
                animeEpisodeBinding!!.nextEpisodeButton.visibility = View.GONE
                animeEpisodeBinding!!.prevEpisodeButton.visibility = View.GONE
            } else if (episodeType.equals(resources.getString(R.string.la_string), ignoreCase = true) || episodeType.contains(resources.getString(R.string.la_string))) {
                animeEpisodeBinding!!.linearAbove.setBackgroundColor(resources.getColor(R.color.red_series_color))
                animeEpisodeBinding!!.linearBelow.setBackgroundColor(resources.getColor(R.color.red_series_color))
            } else if (episodeType.equals(resources.getString(R.string.special_string), ignoreCase = true) || episodeType.contains(resources.getString(R.string.special_string_lower))) {
                animeEpisodeBinding!!.linearAbove.setBackgroundColor(resources.getColor(R.color.orange_series_color))
                animeEpisodeBinding!!.linearBelow.setBackgroundColor(resources.getColor(R.color.orange_series_color))
            } else if (episodeType.equals(resources.getString(R.string.ova_string), ignoreCase = true) || episodeType.contains(resources.getString(R.string.ova_string))) {
                animeEpisodeBinding!!.linearAbove.setBackgroundColor(resources.getColor(R.color.pink_series_color))
                animeEpisodeBinding!!.linearBelow.setBackgroundColor(resources.getColor(R.color.pink_series_color))
            }
            animeEpisodeBinding!!.textAnimeTitleWatch.text = episodeTitle
            videoStreamResultModel.episodeTitle = episodeTitle
            videoStreamResultModel.animeType = episodeType
            videoStreamResultModel.animeStatus = episodeStatus
            videoStreamResultModel.animeThumb = episodeThumb
            getAnimeWatchData(episodeURL)
        } else {
            Log.e("nowURL", "gak ada")
        }
    }

    private fun getAnimeWatchData(animeEpisodeToWatch: String) {
        progressDialog!!.show()
        val afterCut = animeEpisodeToWatch.substring(21)
        nowEpisodeNumber = afterCut.substring(afterCut.indexOf("episode-") + 8)
        if (nowEpisodeNumber!!.contains("-")) {
            nowEpisodeNumber!!.replace("-", ".")
        }
        episodePresenter.getEpisodeToWatchData(animeEpisodeToWatch)
    }

    private fun parseHtmlToViewableContent(result: String) {
        val doc = Jsoup.parse(result)

        //Anime Details URL settings
        val getElementsAnimeDetails = doc.select("a[href^=https://animeindo.to/anime/]")
        if (getElementsAnimeDetails.isEmpty()) {
            Log.e("VideoDetailNull?", "Ya")
        } else {
            val animeDetailsURL = getElementsAnimeDetails.attr("href")
            if (!animeDetailsURL.startsWith("https://animeindo.to/")) {
                Log.e("VideoresultURLError?", "Ya")
                videoStreamResultModel.animeDetailURL = null
            } else {
                videoStreamResultModel.animeDetailURL = animeDetailsURL
            }
        }
        episodeTitle = doc.getElementsByTag("h1").text()
        if (episodeTitle != null) {
            if (episodeTitle!!.contains("Subtitle")) {
                episodeTitle = episodeTitle!!.substring(0, episodeTitle!!.length - 19)
            } else {
                Log.e("CUT?", "OF COURSE NO!")
            }
        }
        animeEpisodeBinding!!.textAnimeTitleWatch.text = episodeTitle
        //get next and prev URL
        val getElementsNextAndPrevEpisode = doc.select("a[href^=https://animeindo.to/]")
        val nextAndPrevURL = ArrayList<String>()
        nextAndPrevURL.clear()
        for (position in getElementsNextAndPrevEpisode.indices) {
            val element = getElementsNextAndPrevEpisode[position]
            val nextandprevurlsingle = element.absUrl("href")
            if (!nextandprevurlsingle.startsWith("https://animeindo.to/anime/") && nextandprevurlsingle.contains("episode")) {
                nextAndPrevURL.add(nextandprevurlsingle)
            }
        }
        Log.e("nextAndPrevURLstring", Gson().toJson(nextAndPrevURL))
        val nextOrPrevURL: String
        val nextOrPrevEpisodeNumber: String
        if (nextAndPrevURL.isEmpty()) {
            animeEpisodeBinding!!.prevEpisodeButton.visibility = View.GONE
            animeEpisodeBinding!!.nextEpisodeButton.visibility = View.GONE
        } else {
            if (nextAndPrevURL.size < 2) {
                nextOrPrevURL = nextAndPrevURL[0]
                nextOrPrevEpisodeNumber = nextOrPrevURL.substring(nextOrPrevURL.indexOf("episode-") + 8)
                if (nextOrPrevEpisodeNumber.contains("-") || nowEpisodeNumber!!.contains("-")) {
                    val nextOrPrevEpisodeNumberCut = nextOrPrevEpisodeNumber.replace("-", ".")
                    val nowEpisodeNumberCut = nowEpisodeNumber!!.replace("-", ".")
                    if (nowEpisodeNumberCut.endsWith(".tamat/")) {
                        val substring = nowEpisodeNumberCut.substring(0, nowEpisodeNumberCut.length - 6)
                        if (java.lang.Double.parseDouble(nextOrPrevEpisodeNumberCut) < java.lang.Double.parseDouble(substring)) {
                            prevURL = nextAndPrevURL[0]
                            nextURL = null
                            animeEpisodeBinding!!.prevEpisodeButton.visibility = View.VISIBLE
                            animeEpisodeBinding!!.nextEpisodeButton.visibility = View.GONE
                        } else if (java.lang.Double.parseDouble(nextOrPrevEpisodeNumberCut) > java.lang.Double.parseDouble(substring)) {
                            prevURL = null
                            nextURL = nextAndPrevURL[0]
                            animeEpisodeBinding!!.prevEpisodeButton.visibility = View.GONE
                            animeEpisodeBinding!!.nextEpisodeButton.visibility = View.VISIBLE
                        }
                    } else if (nowEpisodeNumberCut.endsWith("/")) {
                        if (java.lang.Double.parseDouble(nextOrPrevEpisodeNumberCut) < java.lang.Double.parseDouble(nowEpisodeNumberCut.substring(0, nowEpisodeNumberCut.length - 1))) {
                            prevURL = nextAndPrevURL[0]
                            nextURL = null
                            animeEpisodeBinding!!.prevEpisodeButton.visibility = View.VISIBLE
                            animeEpisodeBinding!!.nextEpisodeButton.visibility = View.GONE
                        } else if (java.lang.Double.parseDouble(nextOrPrevEpisodeNumberCut) > java.lang.Double.parseDouble(nowEpisodeNumberCut.substring(0, nowEpisodeNumberCut.length - 1))) {
                            prevURL = null
                            nextURL = nextAndPrevURL[0]
                            animeEpisodeBinding!!.prevEpisodeButton.visibility = View.GONE
                            animeEpisodeBinding!!.nextEpisodeButton.visibility = View.VISIBLE
                        }
                    } else {
                        if (java.lang.Double.parseDouble(nextOrPrevEpisodeNumberCut) < java.lang.Double.parseDouble(nowEpisodeNumberCut)) {
                            prevURL = nextAndPrevURL[0]
                            nextURL = null
                            animeEpisodeBinding!!.prevEpisodeButton.visibility = View.VISIBLE
                            animeEpisodeBinding!!.nextEpisodeButton.visibility = View.GONE
                        } else if (java.lang.Double.parseDouble(nextOrPrevEpisodeNumberCut) > java.lang.Double.parseDouble(nowEpisodeNumberCut)) {
                            prevURL = null
                            nextURL = nextAndPrevURL[0]
                            animeEpisodeBinding!!.prevEpisodeButton.visibility = View.GONE
                            animeEpisodeBinding!!.nextEpisodeButton.visibility = View.VISIBLE
                        }
                    }
                } else {
                    if (!nowEpisodeNumber!!.endsWith("special/") && !nowEpisodeNumber!!.endsWith("movie/") && !nowEpisodeNumber!!.endsWith("ona/") && !nowEpisodeNumber!!.endsWith("ova/")) {
                        if (nowEpisodeNumber!!.endsWith("/")) {
                            Log.e("WITH SLASH?", "yes")
                            Log.e("URL1", nextOrPrevEpisodeNumber)
                            Log.e("URL2", nowEpisodeNumber!!)
                            if (java.lang.Double.parseDouble(nextOrPrevEpisodeNumber.substring(0, nextOrPrevEpisodeNumber.length - 1)) < java.lang.Double.parseDouble(nowEpisodeNumber!!.substring(0, nowEpisodeNumber!!.length - 1))) {
                                prevURL = nextAndPrevURL[0]
                                nextURL = null
                                animeEpisodeBinding!!.prevEpisodeButton.visibility = View.VISIBLE
                                animeEpisodeBinding!!.nextEpisodeButton.visibility = View.GONE
                            } else if (java.lang.Double.parseDouble(nextOrPrevEpisodeNumber.substring(0, nextOrPrevEpisodeNumber.length - 1)) > java.lang.Double.parseDouble(nowEpisodeNumber!!.substring(0, nowEpisodeNumber!!.length - 1))) {
                                prevURL = null
                                nextURL = nextAndPrevURL[0]
                                animeEpisodeBinding!!.prevEpisodeButton.visibility = View.GONE
                                animeEpisodeBinding!!.nextEpisodeButton.visibility = View.VISIBLE
                            }
                        } else {
                            Log.e("WITHOUT SLASH?", "YES")
                            if (java.lang.Double.parseDouble(nextOrPrevEpisodeNumber) < java.lang.Double.parseDouble(nowEpisodeNumber!!)) {
                                prevURL = nextAndPrevURL[0]
                                nextURL = null
                                animeEpisodeBinding!!.prevEpisodeButton.visibility = View.VISIBLE
                                animeEpisodeBinding!!.nextEpisodeButton.visibility = View.GONE
                            } else if (java.lang.Double.parseDouble(nextOrPrevEpisodeNumber) > java.lang.Double.parseDouble(nowEpisodeNumber!!)) {
                                prevURL = null
                                nextURL = nextAndPrevURL[0]
                                animeEpisodeBinding!!.prevEpisodeButton.visibility = View.GONE
                                animeEpisodeBinding!!.nextEpisodeButton.visibility = View.VISIBLE
                            }
                        }

                    }
                }

            } else if (nextAndPrevURL.size > 2) {
                prevURL = nextAndPrevURL[0]
                nextURL = nextAndPrevURL[1]
                previousEpisodeNumber = prevURL!!.substring(prevURL!!.indexOf("episode-") + 8)
                nextEpisodeNumber = nextURL!!.substring(nextURL!!.indexOf("episode-") + 8)
                animeEpisodeBinding!!.prevEpisodeButton.visibility = View.VISIBLE
                animeEpisodeBinding!!.nextEpisodeButton.visibility = View.VISIBLE
            }
        }


        //Anime videos URL settings
        val getVideoEmbedURL = doc.select("iframe[allowfullscreen=allowfullscreen]")
        val getURLFromElementSrc = getVideoEmbedURL.attr("src")
        val getURLFromElementLazy = getVideoEmbedURL.attr("data-lazy-src")
        val getURLFromElementDataSrc = getVideoEmbedURL.attr("data-src")
        Log.e("VID URL 1", getURLFromElementSrc)
        Log.e("VID URL 2", getURLFromElementLazy)
        Log.e("VID URL 3", getURLFromElementDataSrc)
        Log.e("VID URL 4", getVideoEmbedURL.toString())
        val animeVideoEmbedURL: String
        if (!getURLFromElementSrc.startsWith("//")) {
            animeVideoEmbedURL = if (getURLFromElementLazy.startsWith("https:")) {
                getURLFromElementLazy
            } else {
                if (!getURLFromElementLazy.startsWith("http:")) {
                    "http:$getURLFromElementLazy"
                } else {
                    getURLFromElementLazy
                }
            }
            val animeStreamURL3 = "<html><body style=\"margin: 0; padding: 0\"><iframe width=\"100%\" height=\"100%\" src=\"$animeVideoEmbedURL\" allowfullscreen=\"allowfullscreen\"></iframe></body></html>"
            videoStreamResultModel.videoUrl = animeStreamURL3
            Log.e("allData", Gson().toJson(videoStreamResultModel))
            animeEpisodeBinding!!.webViewWatchAnime.loadData(videoStreamResultModel.videoUrl, "text/html", "utf-8")
        } else if (!getURLFromElementLazy.startsWith("//")) {
            animeVideoEmbedURL = if (getURLFromElementSrc.startsWith("https:")) {
                getURLFromElementSrc
            } else {
                if (!getURLFromElementSrc.startsWith("http:")) {
                    "http:$getURLFromElementSrc"
                } else {
                    getURLFromElementSrc
                }
            }
            val animeStreamURL = "<html><body style=\"margin: 0; padding: 0\"><iframe width=\"100%\" height=\"100%\" src=\"$animeVideoEmbedURL\" allowfullscreen=\"allowfullscreen\"></iframe></body></html>"
            videoStreamResultModel.videoUrl = animeStreamURL
            Log.e("allData", Gson().toJson(videoStreamResultModel))
            animeEpisodeBinding!!.webViewWatchAnime.loadData(videoStreamResultModel.videoUrl, "text/html", "utf-8")
        } else {
            animeVideoEmbedURL = if (getURLFromElementDataSrc.startsWith("https:")) {
                getURLFromElementDataSrc
            } else {
                if (!getURLFromElementDataSrc.startsWith("http:")) {
                    "http:$getURLFromElementDataSrc"
                } else {
                    getURLFromElementDataSrc
                }
            }
            val animeStreamURL2 = "<html><body style=\"margin: 0; padding: 0\"><iframe width=\"100%\" height=\"100%\" src=\"$animeVideoEmbedURL\" allowfullscreen=\"allowfullscreen\"></iframe></body></html>"
            videoStreamResultModel.videoUrl = animeStreamURL2
            Log.e("allData", Gson().toJson(videoStreamResultModel))
            animeEpisodeBinding!!.webViewWatchAnime.loadData(videoStreamResultModel.videoUrl, "text/html", "utf-8")
        }
    }

    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) {
            hideSystemUI()
        }
    }

    public override fun onDestroy() {
        super.onDestroy()
        if (progressDialog != null) {
            progressDialog!!.dismiss()
            progressDialog = null
        }
    }

    private fun hideSystemUI() {
        // Enables regular immersive mode.
        // For "lean back" mode, remove SYSTEM_UI_FLAG_IMMERSIVE.
        // Or for "sticky immersive," replace it with SYSTEM_UI_FLAG_IMMERSIVE_STICKY
        val decorView = window.decorView
        decorView.systemUiVisibility = (View.SYSTEM_UI_FLAG_IMMERSIVE
                // Set the content to appear under the system bars so that the
                // content doesn't resize when the system bars hide and show.
                or View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                // Hide the nav bar and status bar
                or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                or View.SYSTEM_UI_FLAG_FULLSCREEN)
    }

    override fun onGetWatchAnimeEpisodeDataSuccess(watchHTMLResult: String) {
        progressDialog!!.dismiss()
        parseHtmlToViewableContent(watchHTMLResult)
    }

    override fun onGetWatchAnimeEpisodeDataFailed() {
        progressDialog!!.dismiss()
        Toast.makeText(this@WatchAnimeEpisodeActivity, "Your internet connection is worse than your face onii-chan :3", Toast.LENGTH_SHORT).show()
    }
}
